compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp

```diff
@@ -27,6 +30,75 @@ namespace IREE {
 namespace Flow {
 namespace {
 
+// Estimates the evaluation cost of a linalg op using a heuristic cost model.
+static int estimateLinalgOpCost(linalg::LinalgOp op) {
+  if (op.hasDynamicShape()) {
+    // Note: bounded dynamic shapes would be interesting, if the compiler used
+    // them. For now just treat dynamic shapes as arbitrarily large.
+    return INT_MAX;
+  }
+
+  int cost = 1;
+  for (auto loopRange : op.getStaticLoopRanges()) {
+    cost *= loopRange;
+  }
+  LLVM_DEBUG(llvm::dbgs() << "// " << op->getName() << " cost: " << cost
+                          << "\n");
+  return cost;
+}
+
+static std::string loopRangeToString(int64_t loopRange) {
+  return ShapedType::isDynamic(loopRange) ? "?" : llvm::itostr(loopRange);
+}
+
+// Returns a string like "512x?x128" representing a linalg op's loop ranges.
+static std::string getLinalgOpLoopRanges(linalg::LinalgOp op) {
+  auto loopRanges = op.getStaticLoopRanges();
+  std::string outputString;
+  llvm::raw_string_ostream sstream(outputString);
+  llvm::interleave(
+      loopRanges,
+      [&](int64_t loopRange) { sstream << loopRangeToString(loopRange); },
+      [&] { sstream << "x"; });
+  return outputString;
+}
+
+static std::string summarizeLinalgOp(linalg::LinalgOp op) {
+  auto opName = op->getName().getStringRef();
+  if (!opName.consume_front("linalg.")) return "";
+  std::string opSuffix = getLinalgOpLoopRanges(op);
+  // TODO(scotttodd): include element type(s) in this string
+  return opName.str() + (opSuffix.empty() ? "" : "_" + opSuffix);
+}
+
+// Summarizes the contents of a dispatch into a short string.
+// This uses heuristics to aid developer debugging.
+static std::string summarizeDispatchWorkgroupsOp(
+    DispatchWorkgroupsOp regionOp) {
+  Operation *bestOp = NULL;
+  int bestEstimatedCost = -1;
+  std::string bestSummary = "";
+
+  regionOp.getBodyRegion().walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op)
+        .Case<linalg::LinalgOp>([&](auto op) {
+          int estimatedCost = estimateLinalgOpCost(op);
+          if (estimatedCost < bestEstimatedCost) return;
+          bestEstimatedCost = estimatedCost;
+          bestOp = op;
+          bestSummary = summarizeLinalgOp(op);
```

Comment:
We may call this after walking all the ops.

Reply:
We could yeah. This is a little easier to code for when there are multiple cases in the switch.

this way:
```
for op : ops:
  switch(op)
    case linalg:
      if highest cost, pick and summarize
    case linalgext:
      if highest cost, pick and summarize
```

deferred summarization:
```
for op : ops:
  switch(op)
    case linalg:
      if highest cost, pick
    case linalgext:
      if highest cost, pick

switch(highest cost op)
  case linalg:
    summarize
  case linalgext:
    summarize
```

if we introduce an interface for the cost model and summarization, and apply it using an external model, we could just rely on the interface method to handle the polymorphism

Reply:
Okay, I implemented the deferred summarization. Would still be nice to use an interface to avoid the `TypeSwitch` boilerplate.